### DISCUSSION TITLE
Can I detect the backend used by wgpu?

### DISCUSSION THREAD

**Discussion by tseli0s** [parasyte/pixels#339] [Q&A]
Considering this crate uses `wgpu` underneath, is there a way to find out the current backend used by it?

**Comment by parasyte** [comment]
You have a few options!

If you don't need to do anything with the info (you're only interested to satisfy curiosity) you can use the `RUST_LOG` env var (with `env_logger` or `tracing-subscriber`) like this:

```
$ RUST_LOG='wgpu_core::instance=info' cargo run --package minimal-winit
    Finished dev [unoptimized + debuginfo] target(s) in 0.61s
     Running `target/debug/minimal-winit`
[2023-02-26T18:09:14Z INFO  wgpu_core::instance] Adapter Metal AdapterInfo { name: "Intel(R) HD Graphics 630", vendor: 0, device: 0, device_type: IntegratedGpu, driver: "", driver_info: "", backend: Metal }

$ RUST_LOG='wgpu_core::instance=info' WGPU_POWER_PREF=high cargo run --package minimal-winit
    Finished dev [unoptimized + debuginfo] target(s) in 0.57s
     Running `target/debug/minimal-winit`
[2023-02-26T18:10:11Z INFO  wgpu_core::instance] Adapter Metal AdapterInfo { name: "AMD Radeon Pro 560", vendor: 0, device: 0, device_type: DiscreteGpu, driver: "", driver_info: "", backend: Metal }
```

If you want to get access to this at runtime in code, the [`wgpu::Adapter::get_info()`](https://docs.rs/wgpu/latest/wgpu/struct.Adapter.html#method.get_info) API offers some reflection capabilities. But we do not expose this directly through our API. It will need a fork or a design and PR to add some kind of middleware.

The adapter is created here, used for a few things and thrown away: https://github.com/parasyte/pixels/blob/860559272a09b1bf2e2a021b627ea33453d7f029/src/builder.rs#L266-L287

**Comment by tseli0s** [comment]
What I wanted is save it into my game's log if needed so it's a middleground between the two. Maybe I'll make a PR to expose the adapter information through a function, if that's okay with you ofc.

**Reply by parasyte** [reply]
I don't know what a good middleware API for this would look like. You can see the `adapter` is not stored anywhere outside of the builder. We don't need it for anything. Even if the `AdapterInfo` struct is stored, we don't need that either. It would just be holding on to extra state for a method that won't be called by most users.

There are a few other intermediate state items that the builder doesn't keep around. Namely `wgpu::Instance`, but also stuff like `wgpu::SurfaceCapabilities`.

Maybe the builder should accept some `FnMut()` closures that can be executed as "hooks" in the `builder()` method?

```rust
let adapter_info = None;
let pixels = PixelsBuilder::new(width, height, surface_texture)
    .with_adapter_hook(|adapter| {
        *adapter_info = Some(adapter.get_info());
    }
    .build()?;

// Do something with `adapter_info`
if let Some(adapter_info) = adapter_info {
    println!("{adapter_info:?}");
}
```

I don't really love it, but I haven't thought of anything better.
